File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestAnswerComponent.java
package de.westnordost.streetcomplete.quests;

import android.os.Bundle;

import java.util.ArrayList;

import de.westnordost.streetcomplete.data.QuestGroup;

public class QuestAnswerComponent
{
	private static final String ARG_QUEST_ID = "questId";
	private static final String ARG_QUEST_GROUP = "questGroup";

	private OsmQuestAnswerListener callbackListener;

	private long questId;
	private QuestGroup questGroup;

	public static Bundle createArguments(long questId, QuestGroup group)
	{
		Bundle args = new Bundle();
		args.putLong(ARG_QUEST_ID, questId);
		args.putString(ARG_QUEST_GROUP, group.name());
		return args;
	}

	public Bundle getArguments()
	{
		return createArguments(questId, questGroup);
	}

	public void onCreate(Bundle arguments)
	{
		if(arguments == null || arguments.getLong(ARG_QUEST_ID, -1) == -1 ||
				arguments.getString(ARG_QUEST_GROUP, null) == null)
		{
			throw new IllegalStateException("Use QuestAnswerComponent.createArguments and pass the " +
					"created bundle as an argument.");
		}

		questId = arguments.getLong(ARG_QUEST_ID);
		questGroup = QuestGroup.valueOf(arguments.getString(ARG_QUEST_GROUP));
	}

	public void onAttach(OsmQuestAnswerListener listener)
	{
		callbackListener = listener;
	}

	public void onAnswerQuest(Bundle answer)
	{
		callbackListener.onAnsweredQuest(questId, questGroup, answer);
	}

	public void onComposeNote(String questTitle)
	{
		callbackListener.onComposeNote(questId, questGroup, questTitle);
	}

	public void onLeaveNote(String questTitle, String text, ArrayList<String> imagePaths)
	{
		callbackListener.onLeaveNote(questId, questGroup, questTitle, text, imagePaths);
	}

	public void onSkippedQuest()
	{
		callbackListener.onSkippedQuest(questId, questGroup);
	}

	public long getQuestId()
	{
		return questId;
	}

	public QuestGroup getQuestGroup()
	{
		return questGroup;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/QuestAnswerComponent.kt
package de.westnordost.streetcomplete.quests

import android.os.Bundle
import androidx.core.os.bundleOf

import de.westnordost.streetcomplete.data.QuestGroup

class QuestAnswerComponent {

    private lateinit var callbackListener: OsmQuestAnswerListener

    var questId: Long = 0L
        private set
    lateinit var questGroup: QuestGroup
        private set

    val arguments get() = createArguments(questId, questGroup)

    fun onCreate(args: Bundle?) {

        if (args == null || args.getLong(ARG_QUEST_ID, -1) == -1L || args.getString(ARG_QUEST_GROUP, null) == null) {
            throw IllegalStateException(
                "Use QuestAnswerComponent.createArguments and pass the created bundle as an argument."
            )
        }

        questId = args.getLong(ARG_QUEST_ID)
        questGroup = QuestGroup.valueOf(args.getString(ARG_QUEST_GROUP)!!)
    }

    fun onAttach(listener: OsmQuestAnswerListener) {
        callbackListener = listener
    }

    fun onAnsweredQuest(answer: Any) {
        callbackListener.onAnsweredQuest(questId, questGroup, answer)
    }

    fun onComposeNote(questTitle: String) {
        callbackListener.onComposeNote(questId, questGroup, questTitle)
    }

    fun onSplitWay() {
        if (questGroup != QuestGroup.OSM) throw IllegalStateException()
        callbackListener.onSplitWay(questId)
    }

    fun onSkippedQuest() {
        callbackListener.onSkippedQuest(questId, questGroup)
    }

    companion object {
        private const val ARG_QUEST_ID = "questId"
        private const val ARG_QUEST_GROUP = "questGroup"

        fun createArguments(questId: Long, group: QuestGroup) = bundleOf(
            ARG_QUEST_ID to questId,
            ARG_QUEST_GROUP to group.name
        )
    }
}
